### ISSUE TITLE
random DISCONNECTED messages from PhantomJS

### ISSUE THREAD

**Issue by ibawt** [karma-runner/karma#598]
Hi,
  Karma: 0.8.6
  PhantomJS: 1.9
  Node: v0.10.2, v0.10.7

  I'm seeing random disconnect messages in single run mode.  Is this a known problem?  I tried monkey patching the socket.io transports to only do xhr and jsonp but it still happens.  Any ideas?

**Comment by mtscout6** [comment]
I've noticed that using pre-processors can cause this to increase as well. Since karma launches the browser then processes all the files. If your preprocessors take a long time to run then I'll bet you see a direct correlation.

**Comment by nuweb** [comment]
I get the same error.
OSX - 10.9.5
Karma version: 0.12.31
phantomjs version 1.9.8
node --version 0.10.33

->grunt karma
Running "karma:unit" (karma) task
INFO [karma]: Karma v0.12.31 server started at http://localhost:9876/
INFO [launcher]: Starting browser PhantomJS
INFO [PhantomJS 1.9.8 (Mac OS X)]: Connected on socket vo0VCply3aIvM70g5H5C with id 93719535
PhantomJS 1.9.8 (Mac OS X): Executed 1 of 9 SUCCESS (0 secs / 0.001 secs)
PhantomJS 1.9.8 (Mac OS X): Executed 1 of 9 DISCONNECTED (12.386 secs / 0.001 secs)
Warning: Task "karma:unit" failed. Use --force to continue.

Aborted due to warnings.

**Comment by bajtos** [comment]
Adding the following lines to my karma config solved the problem for me:

```diff
     // If browser does not capture in given timeout [ms], kill it
     captureTimeout: 60000,

+    // to avoid DISCONNECTED messages
+    browserDisconnectTimeout : 10000, // default 2000
+    browserDisconnectTolerance : 1, // default 0
+    browserNoActivityTimeout : 60000, //default 10000
```

Source: webpack/karma-webpack#8

**Comment by sloria** [comment]
@bajtos' workaround worked for me.

I am only seeing this error when I run in --single-run mode.

**Comment by veeracs** [comment]
@bajtos workaround did not work for me. I'm consistently getting DISCONNECTED error.

**Comment by anijap** [comment]
@veeracs, I kept getting the same error on a build server with 3.5GB of RAM.
The PhantomJS startup time was consistently around >0.9sec && <=2sec over five builds.

I've increased the *browserNoActivityTimeout* to 3 seconds as a workaround.

**Comment by monitorjbl** [comment]
@bajtos workaround did not work directly for me, I had to increase the retry count to an obscenely high `100` to prevent these random disconnects from breaking the build.

**Comment by Goffen** [comment]
We got the same problem with intermittent test-runs. Once in a while phantomjs wont connect to the websocket in time so the testrunner will timeout and shut down. I think its quite related to the load on the computer running the tests. On everyones developer-machine everything is fine but on the very pressured build-server we get these problems.

**Comment by jeffcai** [comment]
Yes, I got such problem. As Goffen said, it only occurs on the machine with very high load but on local machine it's fine to run test.

**Comment by mik01aj** [comment]
The same for me. I also noticed that this happens mostly with `--single-run`, like @sloria said.

**Comment by blaise-io** [comment]
I also got random disconnects like

```
INFO [launcher]: Starting browser PhantomJS
WARN [launcher]: PhantomJS have not captured in 1000 ms, killing.
INFO [launcher]: Trying to start PhantomJS again (1/2).
WARN [launcher]: PhantomJS have not captured in 1000 ms, killing.
INFO [launcher]: Trying to start PhantomJS again (2/2).
WARN [launcher]: PhantomJS have not captured in 1000 ms, killing.
ERROR [launcher]: PhantomJS failed 2 times (timeout). Giving up.
```

`usePolling: true` in my config solved it.

**Comment by bensquire** [comment]
Same problem here. We run the test in our CI process and this is causing the build to fail. '--single-run' in this case as well.

Interestingly I notice that the first 100 tests run in ~2 seconds, the last 100 (around the 900 mark) take nearly 2 minutes to execute.

@bajtos Settings didn't do anything for us (in fact we already had it set).

**Comment by darrinholst** [comment]
Seeing this as well on our ci server (m3.medium). I traced it down to never reaching [here](https://github.com/karma-runner/karma/blob/master/static/context.html#L33). I then put a bunch of `console.log`s in various spots in our code to see if I could pinpoint a problem with our code, but the results were random. I'm suspicious of the culprit being phantomjs, but I can't figure out how to see what it's doing while it's hanging (top shows that phantomjs is doing _something_). This started happening after upgrading node 0.10 to either iojs 2.5 or iojs 3.3.

    karma 0.12.37
    phantomjs 2.0.1 (debian)
    iojs 2.5 or 3.3

As a temporary workaround I'm going with the following config for karma...

    browserDisconnectTimeout: 5000,
    browserNoActivityTimeout: 5000,
    browserDisconnectTolerance: 10,

This will give it 10 times to try and run. I've found that if it's going to run it's going to start in the first couple seconds hence the 5 second no activity timeout to fail faster.

I can reproduce this reliably if anyone has any suggestions for debugging this.

**Comment by kublanov** [comment]
We also had similar issues in our CI environment.
Karma and PhantomJS did not work properly together, meaning that PhantomJS was able to connect to Karma Server, but did not get any request from it. Also ended up sometimes in:

Disconnected (1 times), because no message in ...

Seems to be related to singleRun mode.

After spending few hours, we've implemented a workaround by using following combination of the two properties:
        browserNoActivityTimeout: 15000
              and
        browserDisconnectTolerance: 5

This way if PhantomJS does not get any messages from Karama Server it tries to reconnect to it, and gets immedeataly served with the messages.

**Comment by gillius** [comment]
This has been happening to us since the beginning of our project about a year and a half ago, but only on our CI server TeamCity. Now we got a newer, nicer, faster server and I hoped all would be better, with it running the newest node and phantomjs. It's not better -- in fact it's a lot worse. It's definitely presenting like a race condition considering that it's random. We're going to try:

```js
captureTimeout: 2000,
browserDisconnectTimeout: 2000,
browserDisconnectTolerance: 3,
browserNoActivityTimeout: 10000,
```
And see if that helps.

**Comment by jdelStrother** [comment]
I started setting `browserNoActivityTimeout: 30000` 3 months ago and haven't had a timeout since.

I believe in our case we'd exceed the default timeout when Chrome starts throttling timers of background tabs.   We use Bluebird promises, and something simple like `Promise.delay(50)` actually kicks off multiple nested timeouts, taking up to 5 seconds to complete when timers are being throttled.

If you're running into the same issue, https://code.google.com/p/chromium/issues/detail?id=485425 might prove useful.

**Comment by gillius** [comment]
In our case, the tests either time out or only take about second to run (there are not many Karma tests). On the old server it would timeout maybe once a week (we do about 20 builds a week), now on the faster server with newer nodejs it happened twice yesterday.

**Comment by gillius** [comment]
It looks like the settings can help:

```
[09:11:48] Running "karma:unit" (karma) task
[09:11:48] INFO [karma]: Karma v0.12.3 server started at http://localhost:9001/
[09:11:48] INFO [launcher]: Starting browser PhantomJS
[09:11:50] WARN [launcher]: PhantomJS have not captured in 2000 ms, killing.
[09:11:50] INFO [launcher]: Trying to start PhantomJS again (1/2).
[09:11:51] INFO [PhantomJS 1.9.8 (Windows 8)]: Connected on socket ebGv2MP9M-QYib0V4pRp with id 67313362
[09:11:52] ..........................................
[09:11:52] PhantomJS 1.9.8 (Windows 8): Executed 42 of 42 [32m SUCCESS [39m (0.162 secs / 0.374 secs)
[09:11:52] 
[09:11:52] Done, without errors.
```

**Comment by moacir-selinger** [comment]
I was having the same problems with the projects where I work. When we have a considerable amount of testing, the problem starts to occur.

As above conversation, I added in my *karma.config.js* the line **browserNoActivityTimeout: 60000** and the problem did not occur anymore.

Thank you for your personal contribution.

**Comment by oshalygin** [comment]
The above configuration from @gillius solved the issue for me.  I think I spent a solid day on this going back and forth.  THe worst part for me was that I was chasing around this error:
 SetProcessDpiAwareness failed: "COM error 0x80070005  (Unknown error 0x0ffffffff80070005)

Which almost(operative word here), always lead to a karma timeout. 

The first few times I saw that the above error and timeout did not ALWAYS happen in unison, I came here.  

For the record the above error is on a Windows Server 2012 CI box.

**Comment by dignifiedquire** [comment]
There is nothing actionable on the karma side of things for this so I'm going to close this.

**Comment by nickawatts** [comment]
I encountered the disconnect problem this week too. After way too much experimentation with different combinations of versions of Node, Karma, Karma launchers and PhantomJS, I realized it had to be something I did, not the tools. I eventually figured out that my problem was a very poorly written (authored by, of course, me) Jasmine spec that uses Angular's `ngMock.$httpBackend ` object to unit test an Angular service. 

I was doing something like this:

```
it('should fail on GET request with invalid username/password', function() {
  $httpBackend.whenGET(/.*/).respond(500);
  ...
  $httpBackend.expectGET("authenticate");
  $httpBackend.flush();
});
```
I'm still not 100% sure why, but this caused something to hang indefinitely, which caused Karma to never respond, which eventually led to PhantomJS timing out. Using `.*` as a regex for possible URIs is a horrible idea that I had forgotten I did "temporarily".

The point is, if you see too have a problem like what is described here, it's likely to be something going wrong in your test suite and not a bug in Karma or PhantomJS.
